### PR TITLE
Bill Splitting page

### LINKED ISSUE #83: Frontend - Bill Splitting
**Is your feature request related to a problem? Please describe.**
Bill splitting is a core feature of the application.

**Describe the solution you'd like**

## These are the specifications, ask for any clarifications

* Add a new PRIVATE page - /bills.
  * Add to private page router.
  * Dashboard: bill splitting card to navigate to bill splitting page.
  * Should list bills:
    * Bills ordered first by not paid, then paid after.
    * Then ordered by the due date (most urgent at top).
    * Suggestion: make design similar to the “upcoming” cards on the dashboard page, but ultimately up to implementers.
    * This card should show the bill name, description, cost.
  * People can click a bill to open a details page.
    * This page should be on /bills/{id} - to implement this and for getting path params, look at the react router.
    * There the user can upload proof (pdf or image).
    * This will be uploaded to [here](https://firebase.google.com/docs/storage/web/upload-files).
    * Generate a uuidv4 string (npm uuid), name the file that, send the generated link to the backend.
    * The details page should also name the bill, show description, cost etc.
    * There should be a button (i.e. a toggle) to indicate it is paid.
      * You do not need to upload a bill in order to select this toggle.
    * The OWNER of the task should be able to edit parameters, such as the description, name.
  * Go back to the dashboard button.
  * At the top of the page there should be an option to create a new bill
    * This can be modal:
      * Bill name.
      * Description.
      * Due date.
      * Total cost.
      * Option for splitting bill evenly (default on)
      * If unchecked, show a table of all the people in the house, the user can enter specific quantities for each user to pay.
  * Also ensure a go back to dashboard option at the top that navigates back to the dashboard.

### Checklist:

- [x] I have double checked that the issue does not currently exist. 
- [x] I have added the appropriate labels (e.g. feature, front-end, back-end).
- [x] I have added the **Platform Project** project

### PR DESCRIPTION
**Pull request by FriedhelmWS**
# Description

Bill splitting page

Fixes/resolves #83 

## Screenshots
Please include any screenshots or media that illustrates the changes made

## Type of change
Please delete options that are not relevant.

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [x] **New feature** (non-breaking change which adds functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **Breaking change** (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist:
Leave blank if not applicable

I have completed these steps when making this pull request:
- [ ] I have assigned my name to the issue
- [ ] I have moved the issue to the **In Progress** column
- [ ] I have labelled the PR appropriately
- [ ] I have assigned myself to the PR

Before opening the PR for review:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have moved the linked issue to the **Review in Progress** column

### CODE REVIEW

File: packages/frontend/src/components/bill/NewBillCard.tsx [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+
+const NewBillCard: React.FC<NewBillCardProps> = () => {
+  const [dueDate, setDueDate] = useState(new Date());
+  const [unixTime, setUnixTime] = useState(0);
+  const [totalCost, setTotalCost] = useState('');
+  const [title, setTitle] = useState('');
+  const [detail, setDetail] = useState('');
+  const [flatmateNum, setFlatmateNum] = useState(6);
+  const [splitCost, setSplitCost] = useState('');
+  const [isEvenlySplit, setIsEvenlySplit] = useState(false);
```
Comment (frasermcc9) [inline packages/frontend/src/components/bill/NewBillCard.tsx]: Would prefer if this was either refactored into a custom hook, or most of these moved into a single useState with an object containing all these keys.

**Review by frasermcc9 — COMMENTED** [review]
All looks pretty clean at the moment, just one thing I noticed when having a quick read

File: packages/frontend/src/components/bill/NewBillCard.tsx [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+    users?.map((user, index) => {
+      let newPayment = {
+        id: user.firebaseId,
+        amount: Number(isEvenlySplit ? splitCost : costHash[user.name]),
+        paid: false,
+      };
+      if (index !== 0) {
+        bill['users'].push(newPayment);
+      }
+      return 0;
```
Comment (Sparye) [inline packages/frontend/src/components/bill/NewBillCard.tsx]: just curious what does this do? return 0

Comment (GlockTian) [reply packages/frontend/src/components/bill/NewBillCard.tsx]: It was a quick fix, map method needs a return statement, but our process doesn't need it to return anything

Comment (j-chad) [reply packages/frontend/src/components/bill/NewBillCard.tsx]: maybe replace with forEach then? map doesn't make sense if you aren't transforming the data

Comment (GlockTian) [reply packages/frontend/src/components/bill/NewBillCard.tsx]: true, I agree, this occur in multiple places as well

Comment (GlockTian) [reply packages/frontend/src/components/bill/NewBillCard.tsx]: done

**Comment by frasermcc9** [comment]
From sync discussion:

CSS fixes
  - name list should always be vertical
  - More padding around text boxes
  - More tabular name list (remove coloured bg, less padding etc)

Should error if splits dont sum to total (use the alert system)
American date

Split Evenly should change to have the same behaviour as if you clicked it twice (i.e. it just sets the values).

Cant put spaces in title or description

Switch location of "complete" button and "detail" button

Upload button should be disabled if a file is not selected.

Show proof

Try to use SWR mutation to refetch when posting.

**Comment by frasermcc9** [comment]
Some bugs but avoidable during demo.
 - Don't use more than 3 people
 - Don't use url join code
 - Avoid quickly clicking on newly created things too fast
 - Avoid editing (works but wont update till refresh)

**Review by frasermcc9 — DISMISSED** [review]
Approving because this needs to be in, some issues but core functions work reasonably well.

**Comment by frasermcc9** [comment]
There's backend changes since we want an endpoint for a single bill. This is to improve optimistic responses.

**Review by frasermcc9 — DISMISSED** [review]

**Review by j-chad — DISMISSED** [review]
LGTM

**Review by j-chad — DISMISSED** [review]
LGTM

**Review by frasermcc9 — DISMISSED** [review]

**Review by j-chad — APPROVED** [review]
LGTM

**Review by frasermcc9 — APPROVED** [review]

**Pull request merged**